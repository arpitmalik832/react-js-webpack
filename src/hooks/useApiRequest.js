import { handleRequest } from '../utils/apiUtils';

const useApiRequest = () => {
  let abortControllers = {};

  const createAbortController = key => {
    if (!abortControllers[key]) {
      abortControllers[key] = new AbortController();
    }

    return abortControllers[key];
  };

  const makeGetCall = ({ axiosInstance, url, config }) => {
    const abortController = createAbortController(JSON.stringify(url));

    const { signal } = abortController;

    return handleRequest(
      axiosInstance.get(url, {
        ...config,
        signal,
      }),
    );
  };

  const makePostCall = ({ axiosInstance, url, config, body }) => {
    const abortController = createAbortController(JSON.stringify(url));

    const { signal } = abortController;

    return handleRequest(
      axiosInstance.post(url, body, {
        ...config,
        signal,
      }),
    );
  };

  const makePutCall = ({ axiosInstance, url, config, body }) => {
    const abortController = createAbortController(JSON.stringify(url));

    const { signal } = abortController;

    return handleRequest(
      axiosInstance.put(url, body, {
        ...config,
        signal,
      }),
    );
  };

  const makeDeleteCall = ({ axiosInstance, url, config }) => {
    const abortController = createAbortController(JSON.stringify(url));

    const { signal } = abortController;

    return handleRequest(
      axiosInstance.delete(url, {
        ...config,
        signal,
      }),
    );
  };

  const cancelRequest = key => {
    if (abortControllers[JSON.stringify(key)]) {
      abortControllers[JSON.stringify(key)].abort();
      delete abortControllers[JSON.stringify(key)];
    }
  };

  const cancelAllRequests = () => {
    Object.keys(abortControllers).forEach(key => {
      abortControllers[key].abort();
    });
    abortControllers = {};
  };

  return {
    makeGetCall,
    makePostCall,
    makePutCall,
    makeDeleteCall,
    cancelRequest,
    cancelAllRequests,
  };
};

export default useApiRequest;
