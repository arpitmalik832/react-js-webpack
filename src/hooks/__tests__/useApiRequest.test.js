/* eslint-disable no-console */
import { renderHook, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import useApiRequest from '../useApiRequest';

jest.mock('../../utils/commonUtils', () => ({
  __esModule: true,
  logRequest: jest.fn(),
  logResponse: jest.fn(),
  errorLogRequest: jest.fn(),
  errorLogResponse: jest.fn(),
}));

describe('useApiRequest unit tests', () => {
  afterEach(cleanup);

  const axiosInstance = {
    interceptors: {
      request: {
        use: (callback, errorCallback) => {
          callback();
          errorCallback();
        },
      },
      response: {
        use: (callback, errorCallback) => {
          callback();
          errorCallback();
        },
      },
    },
    get: (x, y) => console.log(x, y),
    post: (x, y, z) => console.log(x, y, z),
    put: (x, y, z) => console.log(x, y, z),
    delete: (x, y) => console.log(x, y),
  };

  it('snapshot test', () => {
    const hook = renderHook(useApiRequest);

    expect(hook).toMatchSnapshot();
  });

  it('triggering useGetCall function', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makeGetCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
    });
  });

  it('triggering useGetCall function 2 times to test createAbortController else condition', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makeGetCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
    });

    result.current.makeGetCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
    });
  });

  it('triggering usePostCall function', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makePostCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
      body: {
        key: 'value',
      },
    });

    // expect(logRequest).toHaveBeenCalledTimes(1);
    // expect(logResponse).toHaveBeenCalledTimes(1);
  });

  it('triggering usePutCall function', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makePutCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
      body: {
        key: 'value',
      },
    });
  });

  it('triggering useDeleteCall function', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makeDeleteCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
    });
  });

  it('triggering useGetCall function with cancelRequest', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makeGetCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
    });

    result.current.cancelRequest('/todos/1');
    result.current.cancelRequest('/todos/2');
  });

  it('triggering useGetCall function with cancelAllRequests', () => {
    const { result } = renderHook(useApiRequest);

    result.current.makeGetCall({
      axiosInstance,
      url: '/todos/1',
      config: {},
    });

    result.current.cancelAllRequests();
  });
});
