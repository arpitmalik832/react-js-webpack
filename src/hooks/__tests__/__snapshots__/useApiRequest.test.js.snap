// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useApiRequest unit tests snapshot test 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": {
      "cancelAllRequests": [Function],
      "cancelRequest": [Function],
      "makeDeleteCall": [Function],
      "makeGetCall": [Function],
      "makePostCall": [Function],
      "makePutCall": [Function],
    },
  },
  "unmount": [Function],
}
`;
